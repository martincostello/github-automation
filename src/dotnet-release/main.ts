// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as core from '@actions/core';
import { getOctokit } from '@actions/github';
import { getFileContents } from '../shared/github';

/* eslint-disable no-console */

export async function run(): Promise<void> {
  try {
    const token = core.getInput('github-token', { required: false });
    const previousSha = core.getInput('ref', { required: false });
    const github = getOctokit(token);

    const owner = 'dotnet';
    const repo = 'core';

    const { data: branch } = await github.rest.repos.getBranch({
      owner,
      repo,
      branch: 'main',
    });

    const currentSha = branch.commit.sha;

    console.log(` Current SHA: ${currentSha}`);
    console.log(`Previous SHA: ${previousSha}`);

    let releaseNotesFiles: string[] = [];
    let updatedSha = '';

    if (currentSha !== previousSha) {
      const { data: diff } = await github.rest.repos.compareCommitsWithBasehead({
        owner,
        repo,
        basehead: `${previousSha}...${currentSha}`,
      });
      updatedSha = currentSha;
      if (diff.files) {
        releaseNotesFiles = diff.files
          .map((file) => file.filename)
          .filter((file) => file.startsWith('release-notes/') && file.endsWith('/releases.json'));
      }

      console.log(`${releaseNotesFiles} release notes file(s) were updated.`);
    }

    const releaseNotesUpdated = releaseNotesFiles.length > 0;
    const releaseNotes: any[] = [];

    if (releaseNotesUpdated) {
      for (const path of releaseNotesFiles) {
        const release = await getFileContents(github, owner, repo, path, currentSha);

        // Remove the releases array from the release notes
        // otherwise the JSON is too large to use as an output.
        const notes: ReleaseChannel = JSON.parse(release);
        delete notes.releases;

        console.log(`Release notes for ${path}:\n${JSON.stringify(notes, null, 2)}`);

        releaseNotes.push(notes);
      }
    }

    core.setOutput('dotnet-core-updated-sha', updatedSha);
    core.setOutput('dotnet-release-notes', JSON.stringify(releaseNotes));
    core.setOutput('dotnet-releases-updated', releaseNotesUpdated);
  } catch (error: any) {
    core.error(error);
    if (error instanceof Error) {
      core.setFailed(error.message);
    }
  }
}

if (require.main === module) {
  run();
}

interface ReleaseChannel {
  'channel-version': string;
  'latest-release': string;
  'latest-release-date': string;
  'latest-runtime': string;
  'latest-sdk': string;
  'release-type': string;
  'support-phase': string;
  'eol-date"': string;
  'lifecycle-policy"': string;
  'releases'?: Release[];
}

interface Release {
  'release-date': string;
  'release-version': string;
  'security': boolean;
  'release-notes': string;
}
