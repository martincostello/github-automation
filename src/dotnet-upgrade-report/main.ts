// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as core from '@actions/core';
import { getOctokit } from '@actions/github';
import { Context } from '@actions/github/lib/context';
import { fetch } from 'undici';
import { getBadge } from '../shared/badges';
import { ReleasesIndex } from '../shared/dotnet';
import { handle } from '../shared/errors';
import { getDotNetSdk, getFileContents, getPull, getReposForCurrentUser, getWorkflowConfig } from '../shared/github';

export async function run(): Promise<void> {
  try {
    const dotnetNextBranch = 'dotnet-vnext';

    const branch = (core.getInput('branch', { required: false }) || dotnetNextBranch).trim();
    const gist_id = core.getInput('gist-id', { required: false });
    const token = core.getInput('github-token', { required: false });
    const channel = core.getInput('channel', { required: false });

    const context = new Context();
    const github = getOctokit(token);

    const repositories = (await getReposForCurrentUser(github, 'all')).map((repo) => repo.full_name);
    const { checksOfInterest } = await getWorkflowConfig(github, context);

    let latestVersion = '';

    if (branch === dotnetNextBranch || !channel) {
      const releasesIndex = await getFileContents(github, 'dotnet', 'core', 'release-notes/releases-index.json', 'main');
      const releases: ReleasesIndex = JSON.parse(releasesIndex);
      if (releases['releases-index']?.length < 0) {
        throw new Error('No releases found in releases-index.json.');
      }
      latestVersion = releases['releases-index'][0]['latest-sdk'];
    } else {
      const quality = core.getInput('quality', { required: false }) || 'daily';
      const versionsFile = await fetch(`https://aka.ms/dotnet/${channel}/${quality}/productCommit-win-x64.json`);
      const versions = (await versionsFile.json()) as SdkProductCommits;
      if (versions?.sdk?.version) {
        latestVersion = versions.sdk.version.trim();
      }
    }

    const report = ['# .NET vNext Upgrade Report', ''];

    let count = 0;
    let wroteHeader = false;

    for (const slug of repositories) {
      // eslint-disable-next-line no-console
      console.log(`Fetching data for ${slug}.`);

      const [owner, repo] = slug.split('/');
      let commit_sha;
      try {
        const response = await github.rest.repos.getBranch({
          owner,
          repo,
          branch,
        });
        commit_sha = response.data.commit.sha;
      } catch {
        core.debug(`The ${branch} branch of ${slug} does not exist or does not have an open pull request.`);
        continue;
      }

      const dotnetSdk = await getDotNetSdk(github, owner, repo, branch);

      if (!dotnetSdk) {
        core.debug(`The ${branch} branch of ${slug} does not have a global.json file.`);
        continue;
      }

      const {
        data: { default_branch },
      } = await github.rest.repos.get({
        owner,
        repo,
      });

      const baseRefs = [default_branch];

      if (branch !== dotnetNextBranch) {
        baseRefs.push(dotnetNextBranch);
      }

      const { data: prs } = await github.rest.repos.listPullRequestsAssociatedWithCommit({
        owner,
        repo,
        commit_sha,
      });

      if (prs.length < 1) {
        continue;
      }

      const pull_for_ref = prs.find((pull) => baseRefs.includes(pull.base.ref));
      if (!pull_for_ref) {
        core.debug(
          `The ${branch} branch of ${slug} does not have an open pull request targeting any of the following branch(es): ${baseRefs.join(
            ','
          )}.`
        );
        continue;
      }

      const pull = await getPull(github, owner, repo, pull_for_ref.number);
      const hasConflicts = pull.mergeable === false;

      const { data: checkStatuses } = await github.rest.checks.listForRef({
        owner,
        repo,
        ref: commit_sha,
      });

      const statusesOfInterest = checkStatuses.check_runs
        .filter((check) => checksOfInterest.includes(check.name))
        .map((check) => ({
          status: check.status,
          conclusion: check.conclusion,
        }));

      let combinedStatus = 'success';

      for (const status of statusesOfInterest) {
        if (status.conclusion === 'failure') {
          combinedStatus = 'failure';
          break;
        } else if (status.status !== 'completed') {
          combinedStatus = 'pending';
        }
      }

      const sdkVersion = dotnetSdk.version;

      const buildColor = combinedStatus === 'success' ? 'brightgreen' : combinedStatus === 'pending' ? 'yellow' : 'red';
      const buildBadge = getBadge('build', combinedStatus, buildColor, 'github');
      const buildUrl = pull.html_url;

      const purple = '512BD4';
      const sdkColor = sdkVersion === latestVersion ? purple : 'yellow';
      const sdkBadge = getBadge('SDK', sdkVersion, sdkColor, 'dotnet');
      const sdkUrl = `${context.serverUrl}/${slug}/blob/${branch}/global.json#L${dotnetSdk.line}`;

      const conflictsColor = hasConflicts ? 'red' : 'brightgreen';
      const conflictsBadge = getBadge('Conflicts', hasConflicts ? 'Yes' : 'No', conflictsColor, 'git');
      const conflictsUrl = pull.html_url;

      count++;

      if (!wroteHeader) {
        report.push('| Pull Request | SDK Version | Build Status | Conflicts? |');
        report.push('| :----------- | :---------- | :----------- | :--------: |');
        wroteHeader = true;
      }

      report.push(
        `| [${slug}#${pull.number}](${pull.html_url}) | [![.NET SDK version](${sdkBadge})](${sdkUrl}) | [![Build: ${combinedStatus}](${buildBadge})](${buildUrl}) | [![Merge conflicts?](${conflictsBadge})](${conflictsUrl}) |`
      );
    }

    if (count === 0) {
      report.push(`No pull requests found for the \`${branch}\` branch.`);
    }

    report.push('');
    report.push(
      `Generated by [GitHub Actions](${context.serverUrl}/${context.repo.owner}/${context.repo.repo}/actions/runs/${context.runId})`
    );

    const reportText = report.join('\n');

    await core.summary.addRaw(reportText).write();

    if (gist_id) {
      const { data: gist } = await github.rest.gists.update({
        gist_id,
        files: {
          'summary.md': {
            content: reportText,
          },
        },
      });
      if (gist.html_url) {
        core.notice(gist.html_url);
      }
    }
  } catch (error) {
    handle(error);
  }
}

if (require.main === module) {
  run();
}

interface ProductCommit {
  commit: string;
  version: string;
}

interface SdkProductCommits {
  runtime: ProductCommit;
  aspnetcore: ProductCommit;
  windowsdesktop: ProductCommit;
  sdk: ProductCommit;
}
