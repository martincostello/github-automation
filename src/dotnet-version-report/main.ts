// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as core from '@actions/core';
import { getOctokit } from '@actions/github';
import { Context } from '@actions/github/lib/context';
import { getBadge } from '../shared/badges';
import { handle } from '../shared/errors';
import { getFileContents, getDotNetSdk, getReposForCurrentUser } from '../shared/github';
import { ReleaseChannel } from '../shared/dotnet';

/* eslint-disable no-console */

export async function run(): Promise<void> {
  try {
    const token = core.getInput('github-token', { required: false });
    const context = new Context();
    const github = getOctokit(token);

    const repos = await getReposForCurrentUser(github, 'owner');

    const releases: ReleaseChannel[] = JSON.parse(
      await getFileContents(github, 'dotnet', 'core', 'release-notes/releases-index.json', 'main')
    );

    const report = ['# .NET SDK Version Report', '', '| Repository | SDK Version |', '| :--------- | :---------- |'];

    for (const repository of repos) {
      const slug = repository.full_name;
      const branch = repository.default_branch;

      console.log(`Fetching data for ${slug}.`);

      const dotnetSdk = await getDotNetSdk(github, repository.owner, repository.repo, branch);

      if (!dotnetSdk) {
        console.log(`No global.json found in ${slug}.`);
        continue;
      }

      const sdkVersion = dotnetSdk.version;
      const parts = sdkVersion.split('.');
      const channel = `${parts[0]}.${parts[1]}`;

      const [latestVersion] = releases['releases-index'].filter((p) => p['channel-version'] === channel).map((p) => p['latest-sdk']);

      if (!latestVersion) {
        console.info(`No latest version found for channel ${channel}`);
        continue;
      }

      const purple = '512BD4';
      const sdkColor = sdkVersion === latestVersion || latestVersion.startsWith(sdkVersion) ? purple : 'yellow';
      const sdkBadge = getBadge('SDK', sdkVersion, sdkColor, 'dotnet');
      const sdkUrl = `${context.serverUrl}/${slug}/blob/${branch}/global.json#L${dotnetSdk.line}`;

      report.push(`| [${slug}](${repository.html_url}) | [![.NET SDK version](${sdkBadge})](${sdkUrl}) |`);
    }

    await core.summary.addRaw(report.join('\n')).write();
  } catch (error) {
    handle(error);
  }
}

if (require.main === module) {
  run();
}
