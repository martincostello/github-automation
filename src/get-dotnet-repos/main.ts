// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as core from '@actions/core';
import { getOctokit } from '@actions/github';
import { Context } from '@actions/github/lib/context';
import { handle } from '../shared/errors';
import { getDotNetSdk, getReposForCurrentUser } from '../shared/github';

type DotNetRepository = {
  ref: string;
  repo: string;
};

export async function run(): Promise<void> {
  try {
    const branch = core.getInput('branch', { required: false });
    const token = core.getInput('github-token', { required: false });
    const github = getOctokit(token);

    const ref = branch;
    const repositories = await getReposForCurrentUser(github, 'member');

    let singleRepository = core.getInput('repository', { required: false });

    if (singleRepository && !singleRepository.includes('/')) {
      const context = new Context();
      singleRepository = `${context.repo.owner}/${singleRepository}`;
    }

    const result: DotNetRepository[] = [];

    for (const repository of repositories) {
      const full_name = repository.full_name;

      if (singleRepository && singleRepository !== full_name) {
        core.debug(`Skipping repository ${full_name}.`);
        continue;
      }

      const owner = repository.owner;
      const repo = repository.repo;

      try {
        // eslint-disable-next-line no-console
        console.log(`Fetching data for ${full_name}.`);

        if (!(await getDotNetSdk(github, owner, repo, branch))) {
          core.debug(`The ${branch} branch of ${full_name} does not exist or does not have a global.json file.`);
          continue;
        }

        const config: DotNetRepository = {
          ref,
          repo: full_name,
        };

        result.push(config);
      } catch (err) {
        core.debug(`Failed to get data for ${full_name}: ${err}`);
      }
    }

    core.setOutput('repositories', JSON.stringify(result));
  } catch (error) {
    handle(error);
  }
}

if (require.main === module) {
  run();
}
