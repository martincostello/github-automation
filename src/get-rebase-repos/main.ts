// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as core from '@actions/core';
import { getOctokit } from '@actions/github';
import { getPull, getReposForCurrentUser } from '../shared/github';

export async function run(): Promise<void> {
  try {
    const headBranch = core.getInput('branch', { required: true });
    const baseBranch = core.getInput('base', { required: false });
    const force = core.getInput('force', { required: false }) === 'true';
    const specificRepo = core.getInput('repository', { required: false }) || '';
    const token = core.getInput('github-token', { required: false });

    const github = getOctokit(token);

    let repositories: string[];

    if (specificRepo) {
      repositories = [specificRepo];
    } else {
      repositories = (await getReposForCurrentUser({ octokit: github }, 'owner')).map((repo) => repo.full_name);
    }

    const result: string[] = [];
    for (const slug of repositories) {
      // eslint-disable-next-line no-console
      console.log(`Fetching data for ${slug}.`);

      const [owner, repo] = slug.split('/');

      let commit_sha;
      try {
        const { data: branch } = await github.rest.repos.getBranch({
          owner,
          repo,
          branch: headBranch,
        });
        commit_sha = branch.commit.sha;
      } catch (err) {
        core.debug(`Could not get branch ${headBranch}: ${err}`);
        continue;
      }

      const { data: prs } = await github.rest.repos.listPullRequestsAssociatedWithCommit({
        owner,
        repo,
        commit_sha,
      });

      if (prs.length < 1) {
        continue;
      }

      let baseRef = baseBranch;
      if (baseRef === '') {
        const { data: repository } = await github.rest.repos.get({
          owner,
          repo,
        });
        baseRef = repository.default_branch;
      }

      let rebaseBranch = force;
      if (!rebaseBranch) {
        const pull_for_ref = prs.find((pull) => pull.base.ref === baseRef);
        if (!pull_for_ref) {
          core.debug(`No pull request found targeting ${baseRef}.`);
          continue;
        }

        rebaseBranch = (await getPull(github, owner, repo, pull_for_ref.number)).mergeable_state === 'dirty';
        if (rebaseBranch) {
          core.notice(`${slug}:${headBranch} needs rebasing.`);
        }
      }

      if (rebaseBranch) {
        result.push(slug);
      }
    }

    core.setOutput('repositories', JSON.stringify(result));
  } catch (error: any) {
    core.error(error);
    if (error instanceof Error) {
      if (error.stack) {
        core.error(error.stack);
      }
      core.setFailed(error.message);
    }
  }
}

if (require.main === module) {
  run();
}
