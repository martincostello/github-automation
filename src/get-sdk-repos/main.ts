// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as core from '@actions/core';
import { getOctokit } from '@actions/github';
import { Context } from '@actions/github/lib/context';
import { UpdateDotNetSdkConfig } from '../shared/config';
import { handle } from '../shared/errors';
import { getDotNetSdk, getReposForCurrentUser, getUpdateConfiguration } from '../shared/github';

type UpdateConfiguration = {
  'channel': string;
  'exclude-nuget-packages': string | undefined;
  'include-nuget-packages': string | undefined;
  'labels': string;
  'quality': string;
  'ref': string;
  'repo': string;
  'update-nuget-packages': boolean;
};

function getConfiguration(overrides: UpdateDotNetSdkConfig | null): UpdateDotNetSdkConfig {
  const config: UpdateDotNetSdkConfig = {};

  if (overrides) {
    core.debug(`Baseline configuration: ${JSON.stringify(config)}`);
    core.debug(`Repository configuration: ${JSON.stringify(overrides)}`);

    const keys = ['channel', 'exclude-nuget-packages', 'include-nuget-packages', 'labels', 'quality', 'update-nuget-packages'];

    for (const key of keys) {
      const value = overrides[key];
      if (value !== undefined) {
        config[key] = value;
      }
    }

    core.debug(`Merged configuration: ${JSON.stringify(config)}`);
  }

  return config;
}

export async function run(): Promise<void> {
  try {
    const branch = core.getInput('branch', { required: false });
    const token = core.getInput('github-token', { required: false });
    const nightlyChannel = core.getInput('nightly-channel', { required: false });
    const github = getOctokit(token);

    const repositories = await getReposForCurrentUser(github, 'member');

    const excludePackages = '';
    const includePackages = 'Microsoft.AspNetCore.,Microsoft.EntityFrameworkCore.,Microsoft.Extensions.,System.Text.Json';
    const labels = 'dependencies,.NET';
    const ref = branch;
    const channel = branch === 'dotnet-nightly' && nightlyChannel ? nightlyChannel : '';
    const quality = branch === 'dotnet-nightly' ? 'daily' : '';

    let singleRepository = core.getInput('repository', { required: false });

    if (singleRepository && !singleRepository.includes('/')) {
      const context = new Context();
      singleRepository = `${context.repo.owner}/${singleRepository}`;
    }

    const result: UpdateConfiguration[] = [];

    for (const repository of repositories) {
      const full_name = repository.full_name;

      if (singleRepository && singleRepository !== full_name) {
        core.debug(`Skipping repository ${full_name}.`);
        continue;
      }

      const owner = repository.owner;
      const repo = repository.repo;

      const repoConfig = await getUpdateConfiguration(github, owner, repo, ref);
      if (repoConfig?.ignore) {
        core.debug(`Ignoring ${full_name}.`);
        continue;
      }

      try {
        // eslint-disable-next-line no-console
        console.log(`Fetching data for ${full_name}.`);

        if (!(await getDotNetSdk(github, owner, repo, branch))) {
          core.debug(`The ${branch} branch of ${full_name} does not exist or does not have a global.json file.`);
          continue;
        }

        const updateConfig = getConfiguration(repoConfig);

        const valueOrDefault = <T>(value: T | undefined, defaultValue: T): T => {
          return value !== undefined ? value : defaultValue;
        };

        const config: UpdateConfiguration = {
          channel,
          'exclude-nuget-packages': valueOrDefault(updateConfig['exclude-nuget-packages'], excludePackages),
          'include-nuget-packages': valueOrDefault(updateConfig['include-nuget-packages'], includePackages),
          labels,
          quality,
          ref,
          'repo': full_name,
          'update-nuget-packages': valueOrDefault(updateConfig['update-nuget-packages'], true),
        };

        result.push(config);
      } catch (err) {
        core.debug(`Failed to get data for ${full_name}: ${err}`);
      }
    }

    core.setOutput('updates', JSON.stringify(result));
  } catch (error: any) {
    handle(error);
  }
}

if (require.main === module) {
  run();
}
