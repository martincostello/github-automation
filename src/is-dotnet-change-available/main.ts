// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as core from '@actions/core';
import { getOctokit } from '@actions/github';
import { handle } from '../shared/errors';
import { XMLParser } from 'fast-xml-parser';
import { Octokit, getFileContents } from '../shared/github';
import { fetch } from 'undici';

const defaultVersion = '9.0';
const owner = 'dotnet';
const repositoryNames = ['aspnetcore', 'efcore', 'installer', 'runtime', 'sdk'];

function getDependencyGraph(): DependencyGraph {
  const createRepository = (repo: string, dependencies: string[], packageName: string | null = null): ProductRepository => {
    return {
      dependencies,
      full_name: `dotnet/${repo}`,
      owner,
      packageName,
      repo,
      sha: '',
    };
  };

  const runtime = createRepository('runtime', [], 'Microsoft.NETCore.App.Ref');
  const efcore = createRepository('efcore', [runtime.full_name], 'Microsoft.EntityFrameworkCore');
  const aspnetcore = createRepository('aspnetcore', [runtime.full_name, efcore.full_name], 'Microsoft.AspNetCore.App.Ref');
  const sdk = createRepository('sdk', [runtime.full_name, aspnetcore.full_name], 'Microsoft.NET.Sdk');
  const installer = createRepository('installer', [sdk.full_name]);

  return {
    root: installer.full_name,
    nodes: {
      [runtime.full_name]: runtime,
      [efcore.full_name]: efcore,
      [aspnetcore.full_name]: aspnetcore,
      [sdk.full_name]: sdk,
      [installer.full_name]: installer,
    },
  };
}

function getDependencySha(name: string, xml: string): string | null {
  const parser = new XMLParser({ ignoreAttributes: false, attributeNamePrefix: '@' });
  const versionDetails = parser.parse(xml);

  const dependencies = versionDetails?.Dependencies?.ProductDependencies?.Dependency;

  if (dependencies && 'find' in dependencies) {
    const dependency = dependencies.find((element: any) => element['@Name'] === name);
    if (dependency && 'Sha' in dependency) {
      return dependency.Sha;
    }
  }

  return null;
}

async function getLatestSdkVersion(channel: string): Promise<LatestInstallerVersion | null> {
  const quality = 'daily';
  const versionUrl = `https://aka.ms/dotnet/${channel}/${quality}/sdk-productVersion.txt`;

  const init = {
    headers: new Headers([['User-Agent', 'martincostello/github-automation']]),
  };

  let response = await fetch(versionUrl, init);

  if (response.status && response.status >= 400) {
    return null;
  }

  const contentType = response.headers.get('content-type');

  if (!(contentType === 'text/plain' || contentType === 'application/octet-stream')) {
    return null;
  }

  const versionRaw = await response.text();
  const version = versionRaw.trim();

  const platform = 'win-x64';
  const commitsUrl = `https://dotnetbuilds.azureedge.net/public/Sdk/${version}/productCommit-${platform}.json`;

  response = await fetch(commitsUrl, init);

  const commitsJson = await response.json();
  const commits = commitsJson as SdkProductCommits;

  if (!commits) {
    return null;
  }

  return {
    version,
    commits,
  };
}

async function findDependencySha(
  octokit: Octokit,
  root: ProductRepository,
  target: ProductRepository,
  graph: DependencyGraph
): Promise<string | null> {
  if (root.full_name === target.full_name) {
    return root.sha;
  }

  if (!target.packageName) {
    return null;
  }

  const xml = await getFileContents(octokit, owner, root.repo, 'eng/Version.Details.xml', root.sha);
  let sha = getDependencySha(target.packageName, xml);

  if (!sha) {
    for (const name of root.dependencies) {
      const dependency = graph.nodes[name];
      if (!dependency) {
        continue;
      }

      sha = await findDependencySha(octokit, dependency, target, graph);

      if (sha) {
        break;
      }
    }
  }

  return sha;
}

export async function run(): Promise<void> {
  try {
    const repo = core.getInput('repository-name', { required: true });
    const pull_number = Number.parseInt(core.getInput('pull-request', { required: true }), 10);
    const token = core.getInput('github-token', { required: false });

    if (!repositoryNames.includes(repo)) {
      throw new Error(`The ${repo} repository is not supported.`);
    }

    const github = getOctokit(token);

    const { data: pull } = await github.rest.pulls.get({
      owner,
      repo,
      pull_number,
    });

    let isAvailable = false;
    let installerVersion = '';

    if (pull.merged && pull.merge_commit_sha && pull.merged_at) {
      const branch = pull.base.ref;
      const merge_commit_sha = pull.merge_commit_sha;

      const releasePrefix = 'release/';
      const channel = branch.startsWith(releasePrefix) ? branch.slice(releasePrefix.length) : defaultVersion;

      const sdkVersion = await getLatestSdkVersion(channel.slice(0, 3));

      if (!sdkVersion) {
        throw new Error(`The SDK version could not be determined for the ${branch} branch of the ${repo} repository.`);
      }

      const graph = getDependencyGraph();

      let repository: ProductRepository | null = null;
      if (sdkVersion) {
        for (const [, dependency] of Object.entries(graph.nodes)) {
          dependency.sha = sdkVersion?.commits[dependency.repo]?.commit;
          if (dependency.repo === repo) {
            repository = dependency;
          }
        }
      }

      if (!repository) {
        throw new Error(`The ${repo} repository is not supported.`);
      }

      const installer = graph.nodes[graph.root];
      const sha = await findDependencySha(github, installer, repository, graph);

      if (sha) {
        const mergedAt = new Date(pull.merged_at);
        mergedAt.setMinutes(mergedAt.getMinutes() - 1);

        const since = mergedAt.toISOString();

        const commits = await github.paginate(github.rest.repos.listCommits, {
          owner,
          repo,
          sha,
          since,
          per_page: 100,
        });

        if (commits.find((commit) => commit.sha === merge_commit_sha)) {
          isAvailable = true;
          installerVersion = sdkVersion.version;
        }
      }
    }

    let report: string;
    if (isAvailable) {
      report = `The changes from [${owner}/${repo}#${pull_number}](${pull.html_url}) are available in version \`${installerVersion}\` of the .NET SDK.`;
    } else {
      report = `The changes from [${owner}/${repo}#${pull_number}](${pull.html_url}) are not yet available in a daily build of the .NET SDK.`;
    }

    await core.summary.addRaw(report).write();

    core.setOutput('is-available', isAvailable);
    core.setOutput('installer-version', installerVersion);
  } catch (error: any) {
    handle(error);
  }
}

if (require.main === module) {
  run();
}

type ProductRepository = {
  full_name: string;
  owner: string;
  repo: string;
  dependencies: string[];
  packageName: string | null;
  sha: string;
};

type ProductCommit = {
  commit: string;
  version: string;
};

type SdkProductCommits = {
  installer: ProductCommit;
  runtime: ProductCommit;
  aspnetcore: ProductCommit;
  windowsdesktop: ProductCommit;
  sdk: ProductCommit;
};

type DependencyGraph = {
  root: string;
  nodes: Record<string, ProductRepository>;
};

type LatestInstallerVersion = {
  version: string;
  commits: SdkProductCommits;
};
