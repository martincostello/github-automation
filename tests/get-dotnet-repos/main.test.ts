// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as core from '@actions/core';
import { afterAll, beforeAll, describe, expect, test } from '@jest/globals';
import { ActionFixture } from '../ActionFixture';
import { run } from '../../src/get-dotnet-repos/main';
import { setup } from '../fixtures';

describe('get-dotnet-repos', () => {
  describe.each([
    ['multiple', 'main', ''],
    ['multiple', 'main', 'martincostello/website'],
  ])('%s on %s for repository "%s"', (name: string, branch: string, repository: string) => {
    let fixture: ActionFixture;

    beforeAll(async () => {
      await setup(`get-dotnet-repos/${name}`);
      await setup('get-dotnet-repos/member-repos');

      fixture = new ActionFixture(run);
      await fixture.run({
        'branch': branch,
        'github-token': 'fake-token',
        'repository': repository,
      });
    });

    afterAll(async () => {
      await fixture?.destroy();
    });

    test('generates no errors', () => {
      expect(core.error).toHaveBeenCalledTimes(0);
      expect(core.setFailed).toHaveBeenCalledTimes(0);
    });

    test('outputs the correct repositories', () => {
      expect(fixture.getOutput('repositories')).toMatchSnapshot();
    });
  });
});
