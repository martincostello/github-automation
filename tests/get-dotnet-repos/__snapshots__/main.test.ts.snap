// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`get-dotnet-repos multiple on main for repository "" outputs the correct repositories 1`] = `"[{"ref":"main","repo":"martincostello/adventofcode"},{"ref":"main","repo":"martincostello/alexa-london-travel"},{"ref":"main","repo":"martincostello/website"}]"`;

exports[`get-dotnet-repos multiple on main for repository "martincostello/website" outputs the correct repositories 1`] = `"[{"ref":"main","repo":"martincostello/website"}]"`;
